### PR TITLE
mpiext/shortfloat: do not create empty libraries

### PR DESCRIPTION
**Pull request by ggouaillardet**

### CODE REVIEW

**Comment by kawashima-fj** [comment]
@ggouaillardet Thanks. This PR is better than my #6429. :+1: 

I tried to improve the your PR to use the same syntax as `OMPI_MPIEXT_xxx_INCLUDE_MPIFH_IN_USEMPIF08` in `configure.m4` but had no success. In the patch below, `OMPI_MPIEXT_${item}_HAVE_OBJECT` Autoconf macro is not expanded as I intended. I understand the reason. Autoconf is a maze for me... :cry:

```patch
diff --git config/ompi_ext.m4 config/ompi_ext.m4
index f820094a72..f2b7753647 100644
--- a/config/ompi_ext.m4
+++ b/config/ompi_ext.m4
@@ -839,11 +839,12 @@ AC_DEFUN([OMPI_EXT_MAKE_LISTS],[
     # Make the list of libraries
     tmp=
     for item in $2 ; do
-       eval nolib=\$NOLIB_${item}
-       AS_IF([test -z "$nolib"],
-             [eval nolib=\$NOLIB_${item}_$4])
-       AS_IF([test -z "$nolib"],
-             [tmp="$tmp "'$(top_builddir)'"/ompi/mpiext/$item/$3/libmpiext_${item}_$4.la"])
+        m4_ifdef([OMPI_MPIEXT_]$item[_HAVE_OBJECT],
+                 [have_object=OMPI_MPIEXT_${item}_HAVE_OBJECT],
+                 [have_object=1])
+        if test "$have_object" != 0; then
+            tmp="$tmp "'$(top_builddir)'"/ompi/mpiext/$item/$3/libmpiext_${item}_$4.la"
+        fi
     done
     $1_LIBS=$tmp
     AC_SUBST($1_LIBS)
diff --git ompi/mpiext/example/configure.m4 ompi/mpiext/example/configure.m4
index ba58344ebf..5298db0092 100644
--- a/ompi/mpiext/example/configure.m4
+++ b/ompi/mpiext/example/configure.m4
@@ -40,3 +40,12 @@ AC_DEFUN([OMPI_MPIEXT_example_NEED_INIT], [1])
 # By default, mpiext_example_mpifh.h is included in the source file
 # of the mpi_f08_ext module. To disable it, define this macro as 0.
 #AC_DEFUN([OMPI_MPIEXT_example_INCLUDE_MPIFH_IN_USEMPIF08], [0])
+
+# By default, $build_dir/ompi/mpiext/example/c/libmpiext_example_c.la
+# (Libtool archive) is added to the dependency list of libmpi.la.
+# This rule is the same for mpif-h and use-mpi-f08 bindings.
+# However, the ar command of macOS refuses to create an archive file
+# which does not contain any object files. If your extension has no
+# object files, i.e. only header files, define this macro as 0 so that
+# the *.la files are not added to the dependency lists.
+#AC_DEFUN([OMPI_MPIEXT_example_HAVE_OBJECT], [0])
diff --git ompi/mpiext/shortfloat/configure.m4 ompi/mpiext/shortfloat/configure.m4
index e93b5acf8b..dc53178e68 100644
--- a/ompi/mpiext/shortfloat/configure.m4
+++ b/ompi/mpiext/shortfloat/configure.m4
@@ -28,8 +28,7 @@ AC_DEFUN([OMPI_MPIEXT_shortfloat_CONFIG],[
              test "$ENABLE_EXT_ALL" = "1"; } && \
            { test "$ac_cv_type_short_float" = "yes" || \
              test "$ac_cv_type_opal_short_float_t" = "yes"; }],
-          [NOLIB_shortfloat=1
-           $1],
+          [$1],
           [$2])
 
     AS_IF([test "$opal_short_float_type" = "_Float16"],
@@ -44,3 +43,6 @@ AC_DEFUN([OMPI_MPIEXT_shortfloat_CONFIG],[
 # and the mpi_f08_ext module. The former is integer and the latter is
 # type(mpi_datatype).
 AC_DEFUN([OMPI_MPIEXT_shortfloat_INCLUDE_MPIFH_IN_USEMPIF08], [0])
+
+# This extension provides only header files for datatype handles.
+AC_DEFUN([OMPI_MPIEXT_shortfloat_HAVE_OBJECT], [0])
```

**Comment by ggouaillardet** [comment]
@kawashima-fj I think the issue is `m4_ifdef()` is evaluated at `autogen.pl` time, but `$item` is only valid at `configure` time. That is why I chose to mimick `$ENABLE_shortfloat` handling even if `$NOLIB_shortfloat` is known at `autogen.pl` time in this case.

**Comment by kawashima-fj** [comment]
@ggouaillardet I added a commit to write `ompi/mpiext/*/configure.m4` in more natural way.

**Comment by kawashima-fj** [comment]
@ggouaillardet Do you have further change? I think this PR is ready to be merged. Thanks.

**Comment by ggouaillardet** [comment]
@kawashima-fj The PR is good for me too, note I included your commits into mine.

@jsquyres do you want to give this a final review before we merge ?

File: config/ompi_ext.m4 [hunk 1]
```diff
@@ -704,6 +707,13 @@ EOF
              [echo "extern const ompi_mpiext_component_t ompi_mpiext_${component};" >> $outfile.extern
               echo "  &ompi_mpiext_${component}, " >> $outfile.struct])
 
+    m4_ifdef([OMPI_MPIEXT_]$1[_HAVE_OBJECT],
+             [have_object=OMPI_MPIEXT_$1_HAVE_OBJECT],
+             [have_object=1])
+    if test $have_object != 0; then
```
[the remark this answers is not in the document]
Comment (kawashima-fj) [reply config/ompi_ext.m4]: @jsquyres It is guaranteed to be a number. Fixed in the new commit.

**Comment by kawashima-fj** [comment]
@jsquyres @ggouaillardet per-bindings macro may be required sometime. The improvement is not straightforward and I don't have enough time this week. I'll revisit it in the next week or so. If someone has enough time to improve, I give over to him/her.

**Comment by jsquyres** [comment]
I added a comment about one-vs-multiple macros into `mpiext/example/configure.m4` and squashed both this and @kawashima-fj's commit into the first commit (so we're back down to 2 commits).

@ggouaillardet If you are ok with this, feel free to merge.

**Comment by jsquyres** [comment]
@ggouaillardet Are you ok with this amended-and-squashed PR?  I ask only because I don't want to see this one dropped/forgotten again (so that we can compile on MacOS again).

**Pull request merged**